thunder/core/transform_common.py

```diff
@@ -363,6 +364,9 @@ def transform_state_dict_for_submodule(
         """
         return state_dict
 
+    def transform_state_dict(self, model: thunder.ThunderModule, state_dict: dict[str, Any]) -> dict[str, Any]:
```

Comment:
I know it's a bit more tedious, but in order for this to be composable with things like materialization, quantization etc. without using too much memory, would it be possible not to introduce `transform_state_dict` but use the `transform_state_dict_for_submodule`? (also, it would be complicated to interleave the two)

Reply:
How to distinguish load_state_dict from state_dict?

Reply:
Oh right, so to my mind, I think this is the reverse so how about we
- add a `reverse_transform_state_dict_for_submodule` that mimics the `transform_state_dict_for_submodule`  but for the opposite direction,
- add a `original_state_dict()` method calling this with the transforms in reverse order and for all submodules,
- also add a `save_original_state_dict(filename)` method that will actually do the same as `torch.save(self.original_state_dict(), filename)` but does so in an incremental way (i.e. does not materialize the full state dict in memory).

WDYT?